src/lib/shim/src/lib.rs

```diff
@@ -24,9 +34,191 @@ pub fn manager_shmem() -> Option<&'static ManagerShmem> {
     unsafe { bindings::shim_managerSharedMem().as_ref() }
 }
 
+mod global_allow_native_syscalls {
+    use super::*;
+
+    static ALLOW_NATIVE_SYSCALLS: ShimTlsVar<Cell<bool>> = ShimTlsVar::new(|| Cell::new(false));
+
+    pub fn get() -> bool {
+        ALLOW_NATIVE_SYSCALLS.get().get()
+    }
+
+    pub fn swap(new: bool) -> bool {
+        ALLOW_NATIVE_SYSCALLS.get().replace(new)
+    }
+}
+
+// We use a page for a stack guard, and up to another page to page-align the
+// stack guard. We assume 4k pages here but detect at runtime if this is too small.
+const SHIM_SIGNAL_STACK_GUARD_OVERHEAD: usize = 4096 * 2;
+
+mod tls_thread_signal_stack {
+    use super::*;
+
+    static THREAD_SIGNAL_STACK: ShimTlsVar<Cell<*mut core::ffi::c_void>> =
+        ShimTlsVar::new(|| Cell::new(core::ptr::null_mut()));
+
+    // Shouldn't need to make this very large, but needs to be big enough to run the
+    // managed process's signal handlers as well - possibly recursively.
+    //
+    // Stack space that's *never* used shouldn't ever become resident, but an
+    // occasional deep stack could force the pages to be resident ever after.  To
+    // mitigate that, we could consider `madvise(MADV_DONTNEED)` after running
+    // signal handlers, to let the OS reclaim the (now-popped) signal handler stack
+    // frames.
+    const SHIM_SIGNAL_STACK_MIN_USABLE_SIZE: usize = 1024 * 100;
+    const SHIM_SIGNAL_STACK_SIZE: usize =
+        SHIM_SIGNAL_STACK_GUARD_OVERHEAD + SHIM_SIGNAL_STACK_MIN_USABLE_SIZE;
+
+    /// # Safety
+    ///
+    /// See `sigaltstack(2)`
+    pub unsafe fn init() {
```

Comment:
Is there some more-specific safety requirements? For example this function doesn't take any arguments, so it's not clear what from sigaltstack needs to be upheld.